source/portfolio/common/algorithm.cpp

```diff
@@ -0,0 +1,104 @@
+//
+// Created by eduardo on 15/03/2021.
+//
+
+#include "algorithm.h"
+#include <iomanip>
+#include <regex>
+#include <stdexcept>
+#include <vector>
+
+namespace portfolio {
+
+    double string_to_double(std::string_view str) {
+        std::string s(str);
+        std::regex e(R"([+-]{0,1}[0-9]+(\.[0-9]+)?)");
+        if (std::regex_match(s, e)) {
+            std::string::size_type sz; // alias of size_t
+            return std::stod(s, &sz);
+        } else {
+            throw std::runtime_error("Value: " + s +
```

Comment:
O regex ficou bom, mas o lançamento do erro continua deixando a função com a mesma funcionalidade do stod. 

Por se tratar de dado externo, você ainda precisa dar catch neste erro em algum lugar. Não pode assumir que o dado nunca vai ter erro em dado externo. As duas únicas opções é conferir se o dado é válido ou lançar um erro e dar catch nele fora da função.

Entre estas opções, se a intenção é dar catch nesse erro, você poderia simplesmente usar o stod, que já lança erros bem mais específicos de acordo com o problema. Basta colocar um catch na chamada da função e chamar o stod mesmo.

Note também que o regex precisa criar uma máquina de estados para comparar as strings, e isso é algumas centenas ou milhares de vezes mais devagar do que passar pela string comparando. Para comparações simples e que não podemos reaproveitar o regex, fazemos um loop mesmo em vez de criar uma máquina de estado a cada vez. Por esse motivo, usar o próprio stod seria algumas centenas de vezes mais rápido que sua implementação também.

Assim, como está, por estes dois critérios, o stod ainda sai na vantagem. Se a intenção é lançar exceção mesmo, o stod vai ganhar sempre.

O problema aqui com o stod é justamente que o stod lança exceção. Queremos tratar o erro e não queremos dar catch, pois tratar o erro é *muito muito muito* mais barato do que dar catch na exceção, que precisa desenrolar todo o estado da pilha. Só damos catch em situações muito extremas, que finalizariam o programa ou que são totalmente fora de nosso controle, e não em erros possíveis do dia a dia.

Como é uma situação muito ordinária, você pode apenas chamar de fora uma função `is_floating` ou algo assim que te diz se a string contém um número de ponto flutuante. Para isso é só dar um loop da string vendo se os chars `c` são entre `c >= '0'` e `c <= '9'` (ou `std::isdigit`) e abrir uma exceção para um `c == '+' || c == '-'` no início e um `c == '.'` no meio. 

```cpp
bool isNumber(const string& str)
{
    for (char const &c : str) {
        if (std::isdigit(c) == 0) { return false; }
    }
    return true;
}
```

Para tratar os sinais, se quiser, pode usar loop com índices e conferir isso também. Por exemplo `std::isdigit(c) == 0 && !(i == 0 && c == '.')`. Para conferir o ponto, é só estender. Ele pode aparecer em qualquer lugar, mas ele só pode aparecer uma vez e proíbe que venha um sinal depois dele. Pode reorganizar da maneira como quiser para ficar mais legível e de acordo com as demandas da API.

Essa função vai te retornar `true` ou `false` e no seu feed você usa essa informação para decidir se usa esta string ou não. Se não for usar, aí tem que ver o que faz. Se resolver usar porque deu `true`, pode chamar o stod para resolver o problema para você. Se deu false, normalmente vai deixar o resultado nulo, a não ser que tenha como consertar sem esse dado. 

Se quiser, pode também atualizar esta função para evitar exceção mesmo e dentro dela chamar o `is_floating` para ver se realmente pode chamar o stod. Se puder, você chama. Se não puder, você retorna `0.0` ou algo assim para evitar retornar erro. Acaba que esta função não seria tão útil pois ainda precisaria chama o `is_floating` no contexto da função chamadora para ver se o dado está OK, pois usar 0.0 em vez de tratar o problema também seria um erro de lógica, apesar de não ser um erro de compilação. Mas é uma função de conveniência para quando um número inválido de sentinela como o 0.0 não for problemático.

Reply:
A função string_to_double foi excluída e para substituí-la foi criada a função is_floating conforme sugestão. De maneira geral, a nova função valida o formato da string de entrada de forma a se garantir que toda vez que a função std::stod for chamada, não haverão erros de conversão. Com isso, foram feitas alterações em funções de alphavantage_data_feed para que elas retornem valores booleanos para indicar quando elas funcionaram ou falharam, facilitando assim a detecção e tratamento de uma eventual falha em tempo de execução.